File: src/test/resources/org/bau/benchmarks/piDigits.py
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

import sys

def main():
    global acc, den, num
    n = 10000
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    acc, den, num = 0, 1, 1
    i, k = 0, 0
    while i < n:
        k += 1
        next_term(k)
        if num > acc:
            continue
        d = extract_digit(3)
        if d != extract_digit(4):
            continue
        eliminate_digit(d)
        print(chr(ord('0') + d), end = "")
        i += 1
        if i % 10 == 0:
            print("\t:%d" % i)

def extract_digit(nth):
    return (num * nth + acc) // den

def eliminate_digit(d):
    global acc, den, num
    acc -= den * d
    acc *= 10
    num *= 10

def next_term(k):
    global acc, den, num
    k2 = k * 2 + 1
    acc += num * 2
    acc *= k2
    den *= k2
    num *= k

main()

File: src/test/resources/org/bau/benchmarks/piDigits.nim
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

import std/[os, strutils, options]
import bigints

var
    acc = 0.initBigInt
    den = 1.initBigInt
    num = 1.initBigInt
    ten = 10.initBigInt

proc extractDigit(nth: int): int =
    let n = (num * nth.initBigInt + acc) div den
    result = toInt[int](n).get()

proc eliminateDigit(d: int) =
    acc -= den * d.initBigInt
    acc *= ten
    num *= ten

proc nextTerm(k: int) =
    let k2 = initBigInt(k * 2 + 1)
    acc += num.shl(1)
    acc *= k2
    den *= k2
    num *= k.initBigInt

proc main() =
    var n = 10000
    let args = commandLineParams()
    if args.len > 0:
        n = parseInt(args[0])
    var i = 0
    var k = 0
    while i < n:
        inc k
        nextTerm(k)
        if num > acc:
            continue
        let d = extractDigit(3)
        if d != extractDigit(4):
            continue
        eliminateDigit(d)
        stdout.write(chr(ord('0') + d))
        inc i
        if i mod 10 == 0:
            stdout.write("\t:", $i, "\n")
    stdout.write("\n")

when isMainModule:
    main()
